hypothesis-python/src/hypothesis/internal/compat.py

```diff
@@ -188,3 +190,47 @@ def bad_django_TestCase(runner):
         from hypothesis.extra.django._impl import HypothesisTestCase
 
         return not isinstance(runner, HypothesisTestCase)
+
+
+def dataclass_asdict(obj, *, dict_factory=dict):
+    """
+    A vendored variant of dataclasses.asdict. Includes the bugfix for
+    defaultdicts (cpython/32056) for all versions. See also issues/3812.
+
+    This should be removed whenever we drop support for 3.11. We can use the
+    standard dataclasses.asdict after that point.
+    """
+    if not dataclasses._is_dataclass_instance(obj):  # pragma: no cover
+        raise TypeError("asdict() should be called on dataclass instances")
+    return _asdict_inner(obj, dict_factory)
+
+
+def _asdict_inner(obj, dict_factory):
+    if dataclasses._is_dataclass_instance(obj):
+        if dict_factory is dict:
+            return {
+                f.name: _asdict_inner(getattr(obj, f.name), dict)
+                for f in dataclasses.fields(obj)
+            }
+        else:  # pragma: no cover
+            result = []
+            for f in dataclasses.fields(obj):
+                value = _asdict_inner(getattr(obj, f.name), dict_factory)
+                result.append((f.name, value))
+            return dict_factory(result)
```

Comment:
I don't think the perf from a dict comprehension is worth the extra code, so:
```suggestion
        return dict_factory(
            (f.name, _asdict_inner(getattr(obj, f.name), dict_factory))
            for f in dataclasses.fields(obj)
        )
```